shared/helpers/puppeteer.js

```diff
@@ -71,23 +75,33 @@ async function axe(page, overrides = {}) {
  * @template {object} HandlerContext
  * @param {import('puppeteer').Page} page - Puppeteer page object
  * @param {string} componentName - The kebab-cased name of the component
- * @param {MacroRenderOptions} [renderOptions] - Nunjucks macro render options
+ * @param {MacroRenderOptions} renderOptions - Nunjucks macro render options
  * @param {BrowserRenderOptions<HandlerContext>} [browserOptions] - Puppeteer browser render options
  * @returns {Promise<import('puppeteer').Page>} Puppeteer page object
  */
 async function render(page, componentName, renderOptions, browserOptions) {
-  await goTo(page, '/tests/boilerplate')
+  await page.setRequestInterception(true)
 
   const exampleName = renderOptions.fixture?.name ?? 'default'
   const exportName = componentNameToClassName(componentName)
   const selector = `[data-module="govuk-${componentName}"]`
 
-  // Inject rendered HTML into the page
-  await page.$eval(
-    '#slot', // See boilerplate.njk `<div id="slot">`
-    (slot, html) => (slot.innerHTML = html),
-    components.render(componentName, renderOptions)
-  )
+  const route = getComponentURL(componentName, {
+    baseURL: new URL('file://'),
+    exampleName
+  })
+
+  // Handle route and assets
+  const middleware = renderMiddleware({
+    body: renderPreview(componentName, renderOptions),
+    route
+  })
+
+  // Enable middleware
+  page.on('request', middleware)
+
+  // Navigate to component preview
+  await goTo(page, route)
```

Comment:
**question** I follow what's happening, but would have thought that we'd just have to fill the content of the `page` with `page.setContent` and have our assets referenced with `file:` URLs for things to work. Is this dance with URLs and `goTo` so we benefit of the centralised things like bringing Puppeteer to front and waiting for fonts to load?

Reply:
Yeah it's a lot tidier for sure 😊

There was the realisation that we'd want to keep:

1. Events like `pageshow`
2. Ability to reload pages
3. Ability to intercept page reloads
4. Ability to intercept page navigation

We only mock a single page request at the moment but there's room to grow

It also means page navigations can go back to the Review app again quite easily:

```patch
const route = getComponentURL(componentName, {
- baseURL: new URL('file://'),
  exampleName
})
```

☝️ But the middleware will need turning off too

Reply:
Cool! Makes sense! And good to know that it's not much to get back to hitting the review app if we really need for some tests 😊